File: esn_tutorial_py/esn_tutorial_py/orchestrator.py
import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient

from std_msgs.msg import Bool
from geometry_msgs.msg import PoseStamped

from esn_msgs.srv import DetectObject
from esn_msgs.action import PickObject

import threading
import time


class Orchestrator(Node):
    def __init__(self):
        super().__init__("orchestrator_node")

        # Dichiarazione parametri
        self.declare_parameter("topic_name", "object_arrived")
        self.declare_parameter("service_name", "detect_object")
        self.declare_parameter("action_name", "take_object")
        self.declare_parameter("object_id", "target_1")

        self.limit_switch_topic = self.get_parameter("topic_name").value
        self.service_name = self.get_parameter("service_name").value
        self.action_name = self.get_parameter("action_name").value
        self.object_id = self.get_parameter("object_id").value

        # Stato interno
        self.busy = False
        self.object_arrived = False

        # Subscriber
        self.create_subscription(Bool, self.limit_switch_topic, self.on_limit_switch, 10)

        # Client servizio
        self.service_client = self.create_client(DetectObject, self.service_name)

        # Client azione
        self.action_client = ActionClient(self, PickObject, self.action_name)

        self.get_logger().info(
            f"Orchestrator ready. Subscribing '{self.limit_switch_topic}', "
            f"service '{self.service_name}', action '{self.action_name}'"
        )

        # Thread di esecuzione
        threading.Thread(target=self.execute, daemon=True).start()

        # self._lock = threading.Lock() In this case not necessary for GIL (Global Interpreter Lock). Take care if list, dict, ...

    # def set_busy(self, value: bool):
    #     with self._lock:
    #         self.busy = value
    # def get_busy(self) -> bool:
    #     with self._lock:
    #         return self.busy


    def on_limit_switch(self, msg: Bool):
        self.object_arrived = msg.data

    def go_pick_the_object(self):
        self.busy = True

        if not self.service_client.wait_for_service(timeout_sec=10.0):
            self.get_logger().error(f"Service '{self.service_name}' not available")
            self.busy = False
            return

        req = DetectObject.Request()
        req.object_id = self.object_id

        future = self.service_client.call_async(req)
        rclpy.spin_until_future_complete(self, future)

        resp = future.result()
        if resp and resp.success:
            self.get_logger().info("VisionSystem: object detected")
            self.send_action_goal(resp.pose)
        else:
            self.get_logger().warn("VisionSystem: object NOT detected")

        self.busy = False

    def send_action_goal(self, pose: PoseStamped) -> bool:
        if not self.action_client.wait_for_server(timeout_sec=10.0):
            self.get_logger().error(f"Action server '{self.action_name}' not available")
            return False

        goal_msg = PickObject.Goal()
        goal_msg.pose = pose

        send_goal_future = self.action_client.send_goal_async(
            goal_msg,
            feedback_callback=self.on_feedback
        )
        rclpy.spin_until_future_complete(self, send_goal_future)

        goal_handle = send_goal_future.result()
        if not goal_handle.accepted:
            self.get_logger().error("Goal was rejected by action server")
            return False

        self.get_logger().info("Goal accepted")

        result_future = goal_handle.get_result_async()
        rclpy.spin_until_future_complete(self, result_future)

        result = result_future.result().result
        code = result_future.result().status

        if code == 4:  # STATUS_SUCCEEDED
            self.get_logger().info(f"Action SUCCEEDED, success={result.success}")
            return result.success
        elif code == 6:  # STATUS_ABORTED
            self.get_logger().warn("Action ABORTED")
        elif code == 2:  # STATUS_CANCELED
            self.get_logger().warn("Action CANCELED")
        else:
            self.get_logger().error("Unknown result code")
        return False

    def on_feedback(self, feedback_msg):
        self.get_logger().info(f"Progress: {feedback_msg.feedback.progress * 100.0:.0f}%")

    def execute(self):
        rate_hz = 10.0
        while rclpy.ok():
            if not self.busy and self.object_arrived:
                self.get_logger().info("-----------------------------")
                self.get_logger().info("New object arrived")
                self.go_pick_the_object()
            time.sleep(1.0 / rate_hz)

        self.get_logger().info("Stopping the orchestrator...")

File: esn_tutorial/src/orchestrator.cpp
#include <esn_tutorial/orchestrator.hpp>

namespace esn
{

Orchestrator::Orchestrator() : rclcpp::Node("orchestrator_node")
{
    this->declare_parameter<std::string>("topic_name", "object_arrived");
    this->limit_switch_topic_ = this->get_parameter("topic_name").as_string();

    this->declare_parameter<std::string>("service_name", "detect_object");
    this->service_name_ = this->get_parameter("service_name").as_string();

    this->declare_parameter<std::string>("action_name", "take_object");
    this->action_name_ = this->get_parameter("action_name").as_string();

    this->declare_parameter<std::string>("object_id", "target_1");
    this->object_id_ = this->get_parameter("object_id").as_string();

    this->limit_switch_sub_ = this->create_subscription<std_msgs::msg::Bool>(
        this->limit_switch_topic_, 10,
        std::bind(&Orchestrator::on_limit_switch, this, std::placeholders::_1));

    this->service_client_ = this->create_client<DetectObject>(this->service_name_);

    this->action_client_ = rclcpp_action::create_client<PickObject>(this, this->action_name_);

    this->busy_ = false;
    this->object_arrived_ = false;

    RCLCPP_INFO(this->get_logger(),
                "Orchestrator ready. Subscribing '%s', service '%s', action '%s'",
                this->limit_switch_topic_.c_str(),
                this->service_name_.c_str(),
                this->action_name_.c_str());

    std::thread(&Orchestrator::execute, this).detach();
}

void Orchestrator::on_limit_switch(const std_msgs::msg::Bool::SharedPtr msg)
{
    this->object_arrived_ = msg->data;
}

void Orchestrator::go_pick_the_object()
{
    this->busy_ = true;

    geometry_msgs::msg::PoseStamped pose;

    if (!this->service_client_->wait_for_service(std::chrono::seconds(10)))
    {
        RCLCPP_ERROR(this->get_logger(), "Service '%s' not available", this->service_name_.c_str());
        this->busy_ = false;
        return;
    }

    auto req = std::make_shared<DetectObject::Request>();
    req->object_id = this->object_id_;

    auto future = this->service_client_->async_send_request(req);

    future.wait();

    auto resp = future.get();
    if (resp->success)
    {
        RCLCPP_INFO(this->get_logger(), "VisionSystem: object detected");
        this->send_action_goal(resp->pose);
    }
    else
    {
        RCLCPP_WARN(this->get_logger(), "VisionSystem: object NOT detected");
    }

    this->busy_ = false;

    return;
}

bool Orchestrator::send_action_goal(const geometry_msgs::msg::PoseStamped & pose)
{
    using rclcpp_action::ResultCode;

    if (!this->action_client_->wait_for_action_server(std::chrono::seconds(10)))
    {
        RCLCPP_ERROR(this->get_logger(), "Action server '%s' not available", this->action_name_.c_str());
        return false;
    }

    PickObject::Goal goal;
    goal.pose = pose;

    rclcpp_action::Client<PickObject>::SendGoalOptions opts;
    opts.feedback_callback      = std::bind(&Orchestrator::on_feedback, this,
                                       std::placeholders::_1, std::placeholders::_2);

    // (opts.goal_response_callback non serve: la risposta al goal viene gestita attendendo gh_future.wait())
    // (opts.result_callback non serve: aspettiamo il risultato con .wait())

    auto gh_future = this->action_client_->async_send_goal(goal, opts);
    gh_future.wait();

    auto gh = gh_future.get();
    if (!gh) {
        RCLCPP_ERROR(this->get_logger(), "Goal was rejected by action server");
        return false;
    }
    RCLCPP_INFO(this->get_logger(), "Goal accepted");

    auto result_future = this->action_client_->async_get_result(gh);
    result_future.wait();

    auto result = result_future.get();
    switch (result.code)
    {
    case rclcpp_action::ResultCode::SUCCEEDED:
        RCLCPP_INFO(this->get_logger(), "Action SUCCEEDED, success=%s",
                    result.result->success ? "true" : "false");
        return result.result->success;
    case rclcpp_action::ResultCode::ABORTED:
        RCLCPP_WARN(this->get_logger(), "Action ABORTED");
        return false;
    case rclcpp_action::ResultCode::CANCELED:
        RCLCPP_WARN(this->get_logger(), "Action CANCELED");
        return false;
    default:
        RCLCPP_ERROR(this->get_logger(), "Unknown result code");
        return false;
    }
}

void Orchestrator::on_feedback(GoalHandle::SharedPtr,
                               const std::shared_ptr<const PickObject::Feedback> feedback)
{
    RCLCPP_INFO(this->get_logger(), "Progress: %.0f%%", feedback->progress * 100.0);
}

void Orchestrator::execute()
{
    rclcpp::Rate rate(10.0);
    while (rclcpp::ok())
    {
        if (!this->busy_ && this->object_arrived_)
        {
            RCLCPP_INFO(this->get_logger(),"-----------------------------");
            RCLCPP_INFO(this->get_logger(),"New object arrived");
            this->go_pick_the_object();
        }
        rate.sleep();
    }

    RCLCPP_INFO(this->get_logger(),"Stopping the orchestrator...");
}


} // namespace esn
